File: benchmarks/clbg/c/fannkuchredux.c
/* The Computer Language Benchmarks Game
   https://salsa.debian.org/benchmarksgame-team/benchmarksgame/

   contributed by Ledrug
*/

#include <stdio.h>
#include <stdlib.h>

int max_flips = 0;
int checksum = 0;

void fannkuch(int n) {
    int perm[n], perm1[n], count[n];
    int r, flips, i, k, tmp;

    for (i = 0; i < n; i++) perm1[i] = i;

    r = n;
    int nperm = 0;

    while (1) {
        while (r != 1) {
            count[r - 1] = r;
            r--;
        }

        for (i = 0; i < n; i++) perm[i] = perm1[i];

        flips = 0;
        while ((k = perm[0]) != 0) {
            int k2 = (k + 1) >> 1;
            for (i = 0; i < k2; i++) {
                tmp = perm[i];
                perm[i] = perm[k - i];
                perm[k - i] = tmp;
            }
            flips++;
        }

        if (flips > max_flips) max_flips = flips;
        checksum += (nperm & 1) ? -flips : flips;
        nperm++;

        while (1) {
            if (r == n) return;
            int perm0 = perm1[0];
            for (i = 0; i < r; i++) perm1[i] = perm1[i + 1];
            perm1[r] = perm0;
            if (--count[r] > 0) break;
            r++;
        }
    }
}

int main(int argc, char *argv[]) {
    int n = argc > 1 ? atoi(argv[1]) : 7;
    fannkuch(n);
    printf("%d\nPfannkuchen(%d) = %d\n", checksum, n, max_flips);
    return 0;
}

File: benchmarks/clbg/rust/fannkuchredux.rs
// The Computer Language Benchmarks Game
// https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
//
// contributed by the Rust Project Developers

fn fannkuch(n: usize) -> (i32, i32) {
    let mut perm = (0..n).collect::<Vec<_>>();
    let mut perm1 = perm.clone();
    let mut count = vec![0; n];
    let mut max_flips = 0;
    let mut checksum = 0;
    let mut r = n;
    let mut nperm = 0;

    loop {
        while r != 1 {
            count[r - 1] = r;
            r -= 1;
        }

        perm.copy_from_slice(&perm1);

        let mut flips = 0;
        while perm[0] != 0 {
            let k = perm[0];
            perm[0..=k].reverse();
            flips += 1;
        }

        if flips > max_flips {
            max_flips = flips;
        }
        checksum += if nperm & 1 == 0 { flips } else { -flips };
        nperm += 1;

        loop {
            if r == n {
                return (checksum, max_flips);
            }
            let perm0 = perm1[0];
            for i in 0..r {
                perm1[i] = perm1[i + 1];
            }
            perm1[r] = perm0;
            count[r] -= 1;
            if count[r] > 0 {
                break;
            }
            r += 1;
        }
    }
}

fn main() {
    let n: usize = std::env::args()
        .nth(1)
        .and_then(|s| s.parse().ok())
        .unwrap_or(7);

    let (checksum, max_flips) = fannkuch(n);
    println!("{}", checksum);
    println!("Pfannkuchen({}) = {}", n, max_flips);
}
